jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdCredentials.java

```diff
@@ -0,0 +1,192 @@
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class OpenIdCredentials
+{
+    private static final Logger LOG = Log.getLogger(OpenIdCredentials.class);
+
+    private final String redirectUri;
+    private final OpenIdConfiguration configuration;
+    private String authCode;
+    private Map<String, Object> response;
+    private Map<String, Object> claims;
+
+    public OpenIdCredentials(String authCode, String redirectUri, OpenIdConfiguration configuration)
+    {
+        this.authCode = authCode;
+        this.redirectUri = redirectUri;
+        this.configuration = configuration;
+    }
+
+    public String getUserId()
+    {
+        return (String)claims.get("sub");
+    }
+
+    public Map<String, Object> getClaims()
+    {
+        return claims;
+    }
+
+    public Map<String, Object> getResponse()
+    {
+        return response;
+    }
+
+    public void redeemAuthCode() throws IOException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("redeemAuthCode() {}", this);
+
+        if (authCode != null)
+        {
+            try
+            {
+                response = claimAuthCode(authCode);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("response: {}", response);
+
+                String idToken = (String)response.get("id_token");
+                if (idToken == null)
+                    throw new IllegalArgumentException("no id_token");
+
+                String accessToken = (String)response.get("access_token");
+                if (accessToken == null)
+                    throw new IllegalArgumentException("no access_token");
+
+                String tokenType = (String)response.get("token_type");
+                if (!"Bearer".equalsIgnoreCase(tokenType))
+                    throw new IllegalArgumentException("invalid token_type");
+
+                claims = decodeJWT(idToken);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("userClaims {}", claims);
+            }
+            finally
+            {
+                // reset authCode as it can only be used once
+                authCode = null;
+            }
+        }
+    }
+
+    public boolean validate()
+    {
+        if (authCode != null)
+            return false;
+
+        // The aud (audience) Claim MUST contain the client_id value.
+        String audience = (String)claims.get("aud");
+        if (!configuration.getClientId().equals(audience))
+        {
+            LOG.warn("Audience claim MUST contain the value of the Issuer Identifier for the OP");
+            return false;
+        }
+
+        // TODO: this does not work for microsoft
+        // Issuer Identifier for the OpenID Provider MUST exactly match the value of the iss (issuer) Claim.
+        String issuer = (String)claims.get("iss");
+        if (!configuration.getOpenIdProvider().equals(issuer))
+        {
+            //LOG.warn("");
+            //return false;
```

Comment:
Can you remove the commented out code, or uncomment it?

Reply:
done, handling this case properly now